contracts/src/core/EigenDADisperserRegistry.sol

```diff
@@ -22,11 +25,65 @@ contract EigenDADisperserRegistry is OwnableUpgradeable, EigenDADisperserRegistr
         _transferOwnership(_initialOwner);
     }
 
-    function setDisperserInfo(uint32 _disperserKey, DisperserInfo memory _disperserInfo) external onlyOwner {
+    /**
+     * @notice Register or update disperser info.
+     * @dev If the key is new, the sender must supply exactly REGISTRATION_DEPOSIT.
+     *      If already registered, only the original registrant may update and must supply an additional REGISTRATION_DEPOSIT.
+     *      All users, including the owner, must pay the full deposit.
+     */
+    function setDisperserInfo(uint32 _disperserKey, DisperserInfo memory _disperserInfo) external payable {
+        bool isOwner = msg.sender == owner();
+        
+        if (disperserRegistrants[_disperserKey] == address(0)) {
+            // New registration - all users must pay deposit
+            require(msg.value == REGISTRATION_DEPOSIT, "Deposit of 0.1 ether required for registration");
+            disperserRegistrants[_disperserKey] = msg.sender;
+            disperserDeposits[_disperserKey] = msg.value;
+        } else {
+            // Update – only the original registrant can update
+            require(disperserRegistrants[_disperserKey] == msg.sender, "Caller is not the registrant");
+            
+            // All users must pay additional deposit for updates
+            require(msg.value == REGISTRATION_DEPOSIT, "Additional deposit of 0.1 ether required for update");
+            disperserDeposits[_disperserKey] += msg.value;
+        }
+        
+        // Set disperser info
         disperserKeyToInfo[_disperserKey] = _disperserInfo;
+        
+        // If the sender is the owner, mark this disperser as owner-created
+        if (isOwner) {
+            isOwnerCreatedDisperser[_disperserKey] = true;
+        }
+        
         emit DisperserAdded(_disperserKey, _disperserInfo.disperserAddress);
     }
 
+    /**
+     * @notice Deregister a disperser entry.
+     * @dev The caller must be the original registrant.
+     *      A TAX_PERCENTAGE of the accumulated deposit is retained; the rest is refunded.
+     *      All users, including the owner, pay the tax.
+     */
+    function deregisterDisperser(uint32 _disperserKey) external {
+        require(disperserRegistrants[_disperserKey] == msg.sender, "Caller is not the registrant");
+        
+        uint256 totalDeposit = disperserDeposits[_disperserKey];
+        require(totalDeposit > 0, "No deposit found");
+        
+        uint256 tax = (totalDeposit * TAX_PERCENTAGE) / 100;
+        uint256 refund = totalDeposit - tax;
+
+        // Clear registration data
+        delete disperserRegistrants[_disperserKey];
+        delete disperserDeposits[_disperserKey];
+        delete disperserKeyToInfo[_disperserKey];
+        delete isOwnerCreatedDisperser[_disperserKey];
+
+        emit DisperserRemoved(_disperserKey, msg.sender);
+        payable(msg.sender).transfer(refund);
```

Comment:
Do you not want to keep a running total of the tax accrued so that the owner can transfer these funds at some point?

Reply:
I assumed that there's balance tracked per contract automatically. I think the funds should be kept in the contract as long as the disperser is registered, and when they deregister they get 1-tax_percentage back. For the taxed funds, I think we can straight-up burn it or send it to rewards

Reply:
I'm not sure you only want to keep track of the balance in the contract, as you need to be able to differentiate between funds that are taxes, and funds that need to be kept around as a refund for each registered disperser.

Reply:
I see, this makes sense to me. I think we should keep the registration amount as long as the disperser is registered (i.e., owner shouldn't transfer funds out, and disperser "stake" is not tax until they deregister. Perhaps we can use `fee` instead of `tax` as `tax` might be more sensitive and misleading.) The owner can keep a running total of `fee` accured in storage whenever someone deregister, and can transfer the funds out